### PR TITLE
HDFS file read support

### PR DESCRIPTION
**Pull request by chenxing-xc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

I would like to add HDFS file read support(based on libhdfs3), this delivery include:

hdfs table function
new insert syntax that server could read file from local disk or hdfs file directly
URI of HDFS is represented as "hdfs://namenodeip:namenodeport/path-of-hdfsfile"
INSERT SYNTAX is as below:
INSERT INTO TABLE FORMAT XXX INFILE 'file_name'
The above syntax is similar to INSERT INTO TABLE SELECT * FROM TABLE_FUNCTION(.....), but it
is more convenient as we don't need to remember table schema and put it into TABLE_FUNCTION.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
`INSERT INFILE` is very controversial feature in this PR:

1. It requires special support in client (to read data from local file). It is expected that I can send data from local file with clickhouse-client using this feature (analogy to `SELECT ... INTO OUTFILE`).

2. It requires special check on server side (to avoid security issues when reading local files).

3. The behaviour is completely different if you profile local path or URL.

4. The way how we distinguish URIs and local paths looks too hacky - need double check it.

5. Better to implement `INSERT ... INFILE ... FORMAT ...` instead of `INSERT ... FORMAT ... INFILE` due to compatibility issues.

6. I cannot file support for `INSERT ... INFILE` in MySQL or Postgres. MySQL uses `LOAD DATA [LOCAL] INFILE` and Postgres (and most other DBMS) uses `COPY`. I also cannot quickly find any DBMS that implement `INSERT ... INFILE`.

7. For our team members, `INSERT ... INFILE` itself sounds controversial. People may think that it will insert data into file, but actually it reads data from file.

**Comment by alexey-milovidov** [comment]
We can proceed first without `INSERT INFILE` feature.

**Comment by chenxing-xc** [comment]
> We can proceed first without `INSERT INFILE` feature.

Ok, I can withdraw this feature, and put new submodules into ClickHouse-Extras. The original goal of "INFILE' feature is to simulate COPY syntax in PG.
BTW, this feature is heavily used in my team because our ETL task is scheduled on a platform without ClickHouse, and server side data access is needed.

**Comment by alexey-milovidov** [comment]
Could you please add an integration test?

If you have some troubles, you can instead write step-by-step instruction on how this feature can be tested (roll our HDFS in Docker, write files into it, and read them with ClickHouse).

**Comment by alexey-milovidov** [comment]
> BTW, this feature is heavily used in my team because our ETL task is scheduled on a platform without ClickHouse, and server side data access is needed.

We should think how we can made this feature more consistent.

**Comment by chenxing-xc** [comment]
> Could you please add an integration test?
> 
> If you have some troubles, you can instead write step-by-step instruction right here on how this feature can be tested (roll our HDFS in Docker, write files into it, and read them with ClickHouse).

1. This feature depends on HDFS cluster, I am not sure how to add regression test in public.
To test it, we can do the following steps:
1) create a table
2) insert some data
3) select it to outfiles
4) put outfiles into hdfs
5) create another table and populate it with files in hdfs by the new syntax. e.g.
    select * from hdfs('url', format, column-description)
    insert into new_table select * from hdfs('url', format, column-description)

**Comment by zhang2014** [comment]
@chenxing-xc  Maybe:
1. INSERT INTO TABLE FUNCTION hdfs('url', format, column-description)
2. SELECT * FROM hdfs('url', format, column-description)

But this needs implement `write` method in StorageHDFS.

**Comment by alexey-milovidov** [comment]
None of the builds succeed.

**Comment by chenxing-xc** [comment]
> None of the builds succeed.

What 's the build error? there are some extra dependency mentioned in submodule libhdfs3 which need to be installed( by apt-get).
e.g.  libxml2, krb5, libuuid, libgsasl

**Comment by alexey-milovidov** [comment]
https://clickhouse-builds.s3.yandex.net/3617/86f1a18185eb1a67e1c0d445352458838fa02804/report.html

**Comment by chenxing-xc** [comment]
> https://clickhouse-builds.s3.yandex.net/3617/86f1a18185eb1a67e1c0d445352458838fa02804/report.html

I saw the following error in cmake output, "CMake Error at /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find CURL (missing: CURL_LIBRARY CURL_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.10/Modules/FindCURL.cmake:48 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  contrib/libhdfs3-cmake/CMakeLists.txt:22 (find_package)" . 

we need to install it

**Comment by alexey-milovidov** [comment]
We have to put all required libraries as contrib submodules. But it's Ok to start testing with system libraries. You can add the required libraries here:
https://github.com/yandex/ClickHouse/blob/master/docker/packager/binary/Dockerfile
for tests to be run.

**Comment by alexey-milovidov** [comment]
```
~/work/libhdfs3/src$ grep -i -r uuid_ .
./rpc/RpcClient.cpp:    uuid_t id;
./rpc/RpcClient.cpp:    uuid_generate(id);
./rpc/RpcClient.cpp:    clientId.resize(sizeof(uuid_t));
./rpc/RpcClient.cpp:    memcpy(&clientId[0], id, sizeof(uuid_t));
./CMakeLists.txt:    TARGET_LINK_LIBRARIES(libhdfs3-static ${LIBUUID_LIBRARIES})
./CMakeLists.txt:    TARGET_LINK_LIBRARIES(libhdfs3-shared ${LIBUUID_LIBRARIES})
./CMakeLists.txt:    INCLUDE_DIRECTORIES(${LIBUUID_INCLUDE_DIRS})
```

libuuid is used to simply generate a value. We can provide our own minimal implementation to avoid extra dependency.

**Comment by alexey-milovidov** [comment]
libcurl is used only for interaction with "Key Management Server" (KMS).
I don't know whether the usage of KMS mandatory.

**Comment by chenxing-xc** [comment]
> ```
> ~/work/libhdfs3/src$ grep -i -r uuid_ .
> ./rpc/RpcClient.cpp:    uuid_t id;
> ./rpc/RpcClient.cpp:    uuid_generate(id);
> ./rpc/RpcClient.cpp:    clientId.resize(sizeof(uuid_t));
> ./rpc/RpcClient.cpp:    memcpy(&clientId[0], id, sizeof(uuid_t));
> ./CMakeLists.txt:    TARGET_LINK_LIBRARIES(libhdfs3-static ${LIBUUID_LIBRARIES})
> ./CMakeLists.txt:    TARGET_LINK_LIBRARIES(libhdfs3-shared ${LIBUUID_LIBRARIES})
> ./CMakeLists.txt:    INCLUDE_DIRECTORIES(${LIBUUID_INCLUDE_DIRS})
> ```
> libuuid is used to simply generate a value. We can provide our own minimal implementation to avoid extra dependency.

how about replace it with boost::uuid lib?

File: dbms/src/Storages/StorageHDFS.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                fuzzyFileNames = uri;
+                uriPrefix.clear();
+            }
+            else
+            {
+                uriPrefix += "/";
+                fuzzyFileNames = uri.substr(uriPrefix.length());
+            }
+
+            std::vector<String> fuzzyNameList = parseDescription(fuzzyFileNames, 0, fuzzyFileNames.length(), ',' , 100/* hard coded max files */);
```
[the remark this answers is not in the document]
Comment (chenxing-xc) [reply dbms/src/Storages/StorageHDFS.cpp]: They are the same :)

**Comment by chenxing-xc** [comment]
> libcurl is used only for interaction with "Key Management Server" (KMS).
> I don't know whether the usage of KMS is mandatory.
KMS is not mandatory, maybe we can tail it or updating HttpClient using Poco::Net lib

File: dbms/src/Storages/StorageHDFS.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply dbms/src/Storages/StorageHDFS.cpp]: Let's leave only the first. Because it is compatible with the syntax of supported "globs" in bash.
And we can eventually add the support for globs in StorageFile.

**Comment by alexey-milovidov** [comment]
> how about replace it with boost::uuid lib?

It's Ok if it doesn't have extra dependencies. We need to just create a wrapper with single function (fake libuuid).

**Comment by alexey-milovidov** [comment]
> KMS is not mandatory, maybe we can tail it or updating HttpClient using Poco::Net lib

It's not really necessarily to rewrite parts of library :)
The only intent is to avoid/postpone full integration of `libcurl` to our build system, if we can start without it.
There are also possible security implications of using `libcurl`, but it's unrelated concern.

**Comment by alexey-milovidov** [comment]
> This feature depends on HDFS cluster, I am not sure how to add regression test in public.

We have to select some HDFS Docker image for integration test... you can save our time if

**Comment by alesapin** [comment]
@chenxing-xc  Unfortunately I can't build it locally:
1) With gcc it doesn't build with error:
```CMake Error at contrib/libhdfs3-cmake/CMake/Platform.cmake:19 (LIST):
  list index: 1 out of range (-1, 0)
Call Stack (most recent call first):
  contrib/libhdfs3-cmake/CMakeLists.txt:17 (include)
```
2) With clang cmake is OK, but I have compilation problems:
```
[57/650] Building CXX object dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o
FAILED: /usr/bin/ccache /usr/bin/clang++-6.0   -DBOOST_SYSTEM_NO_DEPRECATED -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -isystem contrib/jemalloc-cmake/include -isystem contrib/jemalloc-cmake/include_linux_x86_64 -isystem contrib/libsparsehash -isystem contrib/libdivide -isystem contrib/poco/Data/include -Idbms/src -isystem contrib/double-conversion -isystem contrib/libpcg-random/include -Ilibs/libcommon/include -isystem contrib/poco/Foundation/include -Icontrib/zlib-ng -Icontrib/cityhash102/include -isystem contrib/boost -isystem contrib/poco/Net/include -isystem contrib/poco/Util/include -isystem contrib/poco/XML/include -isystem contrib/poco/JSON/include -isystem contrib/re2_st -isystem contrib/re2 -Ilibs/libpocoext/include -Ilibs/libmysqlxx/include -Icontrib/mariadb-connector-c-cmake/linux_x86_64/include -Icontrib/mariadb-connector-c-cmake/common/include -Icontrib/mariadb-connector-c/include -isystem contrib/ssl/include -Icontrib/libbtrie/include -isystem contrib/poco/Data/ODBC/include -Icontrib/unixodbc-cmake/linux_x86_64 -Icontrib/unixodbc/include -Icontrib/unixodbc/libltdl -Icontrib/unixodbc/libltdl/libltdl -isystem contrib/poco/MongoDB/include -isystem contrib/poco/NetSSL_OpenSSL/include -isystem contrib/poco/Crypto/include -isystem contrib/librdkafka/src -fdiagnostics-color=always -std=c++1z  -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall -Wno-unused-command-line-argument  -Wnon-virtual-dtor -Wextra -Wextra-semi -Wcomma -Winconsistent-missing-destructor-override -Wunused-exception-parameter -Wshadow-uncaptured-local -Wredundant-parens -Wzero-as-null-pointer-constant -O2 -g -DNDEBUG -O3    -pthread -MMD -MT dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o -MF dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o.d -o dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o -c dbms/src/Interpreters/InterpreterInsertQuery.cpp
In file included from dbms/src/Interpreters/InterpreterInsertQuery.cpp:27:
dbms/src/IO/ReadBufferFromHDFS.h:4:10: fatal error: 'hdfs/hdfs.h' file not found
#include <hdfs/hdfs.h>
         ^~~~~~~~~~~~~
1 error generated.
[57/650] Building CXX object dbms/CMakeFiles/dbms.dir/src/Storages/StorageHDFS.cpp.o
FAILED: /usr/bin/ccache /usr/bin/clang++-6.0   -DBOOST_SYSTEM_NO_DEPRECATED -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -isystem contrib/jemalloc-cmake/include -isystem contrib/jemalloc-cmake/include_linux_x86_64 -isystem contrib/libsparsehash -isystem contrib/libdivide -isystem contrib/poco/Data/include -Idbms/src -isystem contrib/double-conversion -isystem contrib/libpcg-random/include -Ilibs/libcommon/include -isystem contrib/poco/Foundation/include -Icontrib/zlib-ng -Icontrib/cityhash102/include -isystem contrib/boost -isystem contrib/poco/Net/include -isystem contrib/poco/Util/include -isystem contrib/poco/XML/include -isystem contrib/poco/JSON/include -isystem contrib/re2_st -isystem contrib/re2 -Ilibs/libpocoext/include -Ilibs/libmysqlxx/include -Icontrib/mariadb-connector-c-cmake/linux_x86_64/include -Icontrib/mariadb-connector-c-cmake/common/include -Icontrib/mariadb-connector-c/include -isystem contrib/ssl/include -Icontrib/libbtrie/include -isystem contrib/poco/Data/ODBC/include -Icontrib/unixodbc-cmake/linux_x86_64 -Icontrib/unixodbc/include -Icontrib/unixodbc/libltdl -Icontrib/unixodbc/libltdl/libltdl -isystem contrib/poco/MongoDB/include -isystem contrib/poco/NetSSL_OpenSSL/include -isystem contrib/poco/Crypto/include -isystem contrib/librdkafka/src -fdiagnostics-color=always -std=c++1z  -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall -Wno-unused-command-line-argument  -Wnon-virtual-dtor -Wextra -Wextra-semi -Wcomma -Winconsistent-missing-destructor-override -Wunused-exception-parameter -Wshadow-uncaptured-local -Wredundant-parens -Wzero-as-null-pointer-constant -O2 -g -DNDEBUG -O3    -pthread -MMD -MT dbms/CMakeFiles/dbms.dir/src/Storages/StorageHDFS.cpp.o -MF dbms/CMakeFiles/dbms.dir/src/Storages/StorageHDFS.cpp.o.d -o dbms/CMakeFiles/dbms.dir/src/Storages/StorageHDFS.cpp.o -c dbms/src/Storages/StorageHDFS.cpp
In file included from dbms/src/Storages/StorageHDFS.cpp:7:
dbms/src/IO/ReadBufferFromHDFS.h:4:10: fatal error: 'hdfs/hdfs.h' file not found
#include <hdfs/hdfs.h>
         ^~~~~~~~~~~~~
1 error generated.
```
Seems like some misconfiguration in cmake?

File: cmake/find_llvm.cmake [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Broken in macos. TODO: update clang, re-test, enable
 if (NOT APPLE)
-    option (ENABLE_EMBEDDED_COMPILER "Set to TRUE to enable support for 'compile' option for query execution" 1)
+    option (ENABLE_EMBEDDED_COMPILER "Set to TRUE to enable support for 'compile' option for query execution" 0)
```
Comment (alesapin) [inline cmake/find_llvm.cmake]: Why we need to disable it?

File: dbms/src/Storages/StorageHDFS.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        void readPrefixImpl() override
+        {
+            reader->readPrefix();
+        }
+
+        void readSuffixImpl() override
+        {
+            auto explicitReader = dynamic_cast<UnionBlockInputStream<> *>(reader.get());
```
Comment (alesapin) [inline dbms/src/Storages/StorageHDFS.cpp]: This is uncompilable code. You have to remove empty diamond `<>`.

File: dbms/src/Storages/StorageHDFS.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            if (inputs.size() == 0)
+                throw Exception("StorageHDFS inputs interpreter error", ErrorCodes::BAD_ARGUMENTS);
+
+            if (inputs.size() == 1)
+            {
+                reader = inputs[0];
+            }
+            else
+            {
+                reader = std::make_shared<UnionBlockInputStream<> >(inputs, nullptr, context.getSettingsRef().max_distributed_connections);
```
Comment (alesapin) [inline dbms/src/Storages/StorageHDFS.cpp]: Remove `<>`

File: dbms/CMakeLists.txt [hunk 5]
```diff
@@ -173,6 +173,7 @@ target_link_libraries (clickhouse_common_io
         PRIVATE
     apple_rt
     ${CMAKE_DL_LIBS}
+    ${HDFS3_LIBRARY}
```
Comment (alesapin) [inline dbms/CMakeLists.txt]: Missed `target_include_directories` for HDFS.

File: dbms/src/Storages/StorageHDFS.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        const String & table_name_,
+        const String & format_name_,
+        const ColumnsDescription & columns_,
+        Context & context_);
+
+private:
+    String uri;
+    String format_name;
+    String table_name;
+    Context & context_global;
```
Comment (alesapin) [inline dbms/src/Storages/StorageHDFS.h]:
```
dbms/src/Storages/StorageHDFS.h:49:15: warning: private field 'context_global' is not used [-Wunused-private-field]
    Context & context_global;
```

File: dbms/src/TableFunctions/ITableFunction.h [hunk 7]
```diff
@@ -42,5 +44,6 @@ class ITableFunction
 
 using TableFunctionPtr = std::shared_ptr<ITableFunction>;
 
+std::vector<String> parseDescription(const String & description, size_t l, size_t r, char separator, size_t max_addresses);
```
Comment (alesapin) [inline dbms/src/TableFunctions/ITableFunction.h]: This function doesn't seems too common. Let's move it to separate file?

File: CMakeLists.txt [hunk 8]
```diff
@@ -89,6 +89,7 @@ endif ()
 
 option (TEST_COVERAGE "Enables flags for test coverage" OFF)
 option (ENABLE_TESTS "Enables tests" ON)
+option (ENABLE_INSERT_INFILE "Enables INSERT INFILE syntax" OFF)
```
Comment (alesapin) [inline CMakeLists.txt]: Is it possible to move INSERT INFILE to separate Pull Request?

File: dbms/src/Storages/StorageHDFS.h (hunk 6 above)
Comment (chenxing-xc) [reply dbms/src/Storages/StorageHDFS.h]: Fixed. remove it

File: dbms/CMakeLists.txt (hunk 5 above)
Comment (chenxing-xc) [reply dbms/CMakeLists.txt]: Do you update submodules? libhdfs3 will be in submodule.

File: dbms/src/Storages/StorageHDFS.cpp (hunk 4 above)
Comment (chenxing-xc) [reply dbms/src/Storages/StorageHDFS.cpp]: Fixed

File: dbms/src/Storages/StorageHDFS.cpp (hunk 3 above)
Comment (chenxing-xc) [reply dbms/src/Storages/StorageHDFS.cpp]: Fixed. UnionBlockInputStream was templated, but it is not after rebasing code.

File: cmake/find_llvm.cmake (hunk 2 above)
Comment (chenxing-xc) [reply cmake/find_llvm.cmake]: Fixed. We don't need to disable it. it was turned off in my test env to avoid compiling llvm lib.

File: dbms/CMakeLists.txt (hunk 5 above)
Comment (alesapin) [reply dbms/CMakeLists.txt]: Yes, I have updated submodules, but compiler didn't see `hdfs.h` header without inserting `target_include_directories(dbms ${HDFS_INCLUDE_DIR})` in CMakeLists.txt:
```
[729/1377] Building CXX object dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o
FAILED: /usr/bin/ccache /usr/bin/clang++-6.0   -DBOOST_SYSTEM_NO_DEPRECATED -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -isystem contrib/jemalloc-cmake/include -isystem contrib/jemalloc-cmake/include_linux_x86_64 -isystem contrib/libsparsehash -isystem contrib/libdivide -isystem contrib/poco/Data/include -Idbms/src -isystem contrib/dou
ble-conversion -isystem contrib/libpcg-random/include -Ilibs/libcommon/include -isystem contrib/poco/Foundation/include -Icontrib/zlib-ng -Icontrib/cityhash102/include -isystem contrib/boost -isystem contrib/poco/Net/include -isystem contrib/poco/Util/include -isystem contrib/poco/XML/include -isystem contrib/poco/JSON/include -isystem contrib/re2_st -isystem contrib/re2 -Ilibs/li
bpocoext/include -Ilibs/libmysqlxx/include -Icontrib/mariadb-connector-c-cmake/linux_x86_64/include -Icontrib/mariadb-connector-c-cmake/common/include -Icontrib/mariadb-connector-c/include -isystem contrib/ssl/include -Icontrib/libbtrie/include -isystem contrib/poco/Data/ODBC/include -Icontrib/unixodbc-cmake/linux_x86_64 -Icontrib/unixodbc/include -Icontrib/unixodbc/libltdl -Icont
rib/unixodbc/libltdl/libltdl -isystem contrib/poco/MongoDB/include -isystem contrib/poco/NetSSL_OpenSSL/include -isystem contrib/poco/Crypto/include -isystem contrib/librdkafka/src -fdiagnostics-color=always -std=c++1z  -pipe -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall -Wno-unused-command-line-argument  -Wnon-virtual-dtor -Wextra -Wextra-semi -Wcomma -Winconsistent-
missing-destructor-override -Wunused-exception-parameter -Wshadow-uncaptured-local -Wredundant-parens -Wzero-as-null-pointer-constant -O2 -g -DNDEBUG -O3    -pthread -MMD -MT dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o -MF dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInsertQuery.cpp.o.d -o dbms/CMakeFiles/dbms.dir/src/Interpreters/InterpreterInse
rtQuery.cpp.o -c dbms/src/Interpreters/InterpreterInsertQuery.cpp
In file included from dbms/src/Interpreters/InterpreterInsertQuery.cpp:27:
dbms/src/IO/ReadBufferFromHDFS.h:4:10: fatal error: 'hdfs/hdfs.h' file not found
#include <hdfs/hdfs.h>
         ^~~~~~~~~~~~~
1 error generated.
[729/1377] Building CXX object dbms/CMakeFiles/dbms.dir/src/Interpreters/Join.cpp.o
[729/1377] Building CXX object dbms/CMakeFiles/dbms.dir/src/Interpreters/Aggregator.cpp.o
ninja: build stopped: subcommand failed.
```

**Review by alesapin — COMMENTED** [review]
Also, can you add `libxml2`, `libkrb` and `libgsasl` to contrib or you need help with this libraries?

File: dbms/CMakeLists.txt (hunk 5 above)
Comment (chenxing-xc) [reply dbms/CMakeLists.txt]: I just delivered a fix for this

**Comment by alesapin** [comment]
@chenxing-xc Where did you get libhdfs3? There are a lot of forks on github and we would like to have commits history in our Clickhouse-Extras repo.

**Comment by chenxing-xc** [comment]
> @chenxing-xc Where did you get libhdfs3? There are a lot of forks on github and we would like to have commits history in our Clickhouse-Extras repo.

I got it from [https://github.com/Pivotal-DataFabric/attic-libhdfs3.git]

**Comment by alexey-milovidov** [comment]
@chenxing-xc Congratulations!

**Comment by gubinjie** [comment]
@chenxing-xc  Hello, when we use clickhouse to read hdfs files, the following exception often occurs，Can you tell me what problems may cause this exception, and what methods can be used to resolve this exception，thank you！
-----------------------
executeQuery: Code: 210, e.displayText() = DB::Exception: Fail to read HDFS file: hdfs://myuser@IP::PORT/user/hive
/warehouse/mydb.db/my_table_tsv/000007_0 HdfsIOException: InputStreamImpl: cannot read file: /user/hive/warehouse/mydb.db/mytable_tsv/000007_0, from position 0, size: 1048576. Caused by: HdfsTimeoutException: Read 8 bytes timeout (version 19.16.4.12 (official build)) (from [::ffff:30.18.54.105]:38882) (in query: insert into

。。。

Stack trace:

0x55a39713fed0 StackTrace::StackTrace() /usr/bin/clickhouse
0x55a39713fca5 DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits, std::allocator > const&, int) /usr/bin/clickhouse
0x55a396fb534d ? /usr/bin/clickhouse
0x55a39a9228dd DB::CSVRowInputFormat::readPrefix() /usr/bin/clickhouse
0x55a39ad2642b DB::IRowInputFormat::generate() /usr/bin/clickhouse
0x55a39a8f5dde DB::ISource::work() /usr/bin/clickhouse
0x55a39a8d1385 DB::InputStreamFromInputFormat::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a398db4e98 DB::OwningBlockInputStreamDB::ReadBuffer::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39a6c8f98 ? /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39aa83efb DB::ExpressionBlockInputStream::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39aa83efb DB::ExpressionBlockInputStream::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39a4ca3b2 DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::loop(unsigned long) /usr/bin/clickhouse
0x55a39a4caa75 DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::thread(std::shared_ptrDB::ThreadGroupStatus, unsigned long) /usr/bin/clickhouse
0x55a39a4cb3ed ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::)(std::shared_ptrDB::ThreadGroupStatus, unsigned long), DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler, std::shared_ptrDB::ThreadGroupStatus, unsigned long&>(void (DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::&&)(std::shared_ptrDB::ThreadGroupStatus, unsigned long), DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler&&, std::shared_ptrDB::ThreadGroupStatus&&, unsigned long&)::{lambda()#1}::operator()() const /usr/bin/clickhouse
0x55a39718a47c ThreadPoolImplstd::thread::worker(std::_List_iteratorstd::thread) /usr/bin/clickhouse
0x55a39cea6460 ? /usr/bin/clickhouse
0x7efcdf5ea6ba start_thread /lib/x86_64-linux-gnu/libpthread-2.23.so
0x7efcdef1482d __clone /lib/x86_64-linux-gnu/libc-2.23.so
2020.01.04 05:28:36.088689 [ 379 ] {90a86a21-7ef2-4e29-b16c-fa563a84e528} HTTPHandler: Code: 210, e.displayText() = DB::Exception: Fail to read HDFS file: hdfs://user@IP:PORT/user/hive/warehouse/ mydb.db/my_table_tsv 000007_0 HdfsIOException: InputStreamImpl: cannot read file: /user/hive/warehouse/ mydb.db/my_table_tsv/000007_0, from position 0, size: 1048576. Caused by: HdfsTimeoutException: Read 8 bytes timeout, Stack trace:

0x55a39713fed0 StackTrace::StackTrace() /usr/bin/clickhouse
0x55a39713fca5 DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits, std::allocator > const&, int) /usr/bin/clickhouse
0x55a396fb534d ? /usr/bin/clickhouse
0x55a39a9228dd DB::CSVRowInputFormat::readPrefix() /usr/bin/clickhouse
0x55a39ad2642b DB::IRowInputFormat::generate() /usr/bin/clickhouse
0x55a39a8f5dde DB::ISource::work() /usr/bin/clickhouse
0x55a39a8d1385 DB::InputStreamFromInputFormat::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a398db4e98 DB::OwningBlockInputStreamDB::ReadBuffer::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39a6c8f98 ? /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39aa83efb DB::ExpressionBlockInputStream::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39aa83efb DB::ExpressionBlockInputStream::readImpl() /usr/bin/clickhouse
0x55a39a3523c7 DB::IBlockInputStream::read() /usr/bin/clickhouse
0x55a39a4ca3b2 DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::loop(unsigned long) /usr/bin/clickhouse
0x55a39a4caa75 DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::thread(std::shared_ptrDB::ThreadGroupStatus, unsigned long) /usr/bin/clickhouse
0x55a39a4cb3ed ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::)(std::shared_ptrDB::ThreadGroupStatus, unsigned long), DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler, std::shared_ptrDB::ThreadGroupStatus, unsigned long&>(void (DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler::&&)(std::shared_ptrDB::ThreadGroupStatus, unsigned long), DB::ParallelInputsProcessorDB::UnionBlockInputStream::Handler&&, std::shared_ptrDB::ThreadGroupStatus&&, unsigned long&)::{lambda()#1}::operator()() const /usr/bin/clickhouse
0x55a39718a47c ThreadPoolImplstd::thread::worker(std::_List_iteratorstd::thread) /usr/bin/clickhouse
0x55a39cea6460 ? /usr/bin/clickhouse
0x7efcdf5ea6ba start_thread /lib/x86_64-linux-gnu/libpthread-2.23.so
0x7efcdef1482d __clone /lib/x86_64-linux-gnu/libc-2.23.so
(version 19.16.4.12 (official build))

**Comment by filimonov** [comment]
@gubinjie #9263

**Comment by chenxing-xc** [comment]
this looks like clickhouse cannot access hdfs datanode.  Currently we don't
use the "user" in the uri(user@ip...) to access hdfs, instead use
clickhouse process's user.

filimonov <notifications@github.com> 于2020年2月27日周四 下午1:35写道：

**Comment by gubinjie** [comment]
@chenxing-xc Hello, we want to know what you said using the clickhouse process's user method. If you use this method, will it solve the previous exception? Thank you

**Comment by SenCoder** [comment]
We also get this promblem in clickhouse:

executeQuery: Code: 210, e.displayText() = DB::Exception: Fail to read HDFS file: hdfs://myuser@IP::PORT/xx/xx/xx/xx HdfsIOException: InputStreamImpl: cannot read file: /xx/xx/xx/xx , from position 805306368, size: 1048576. Caused by: HdfsTimeoutException: Read 8 bytes timeout (version 19.16.4.12 (official build)) (from [::ffff:xx.xx.xx.xx]:37456) (in query:
...


I am sure the user has got access to hdfs, and the network is OK.  Is there any configure to change timeout param for reading HDFS ?

thank you~
